python/tvm/relay/quantize/quantize.py

```diff
@@ -187,15 +203,343 @@ def count_conv2d(self):
+
+
+def annotate(graph, layout_map):
+    """Given a float32 graph, annotate will rewrite the graph
+    and return back a graph which simulates the error brought by
+    current quantization scheme.
+
+    Parameters
+    ---------
+    graph: Function
+        The original graph
+
+    Returns
+    -------
+    ret: Function
+        The graph after annotation
+    """
+    _set_conv_counter(0)  # reset counter
+    _set_layout_map(layout_map)
+    return _quantize.annotate(graph)
+
+
+def tag_layout(graph):
+    conv2d_op = _op.get("nn.conv2d")
+    dense_op = _op.get("nn.dense")
+    _op_layout_map = dict()
+    # layouts to tag later
+    deferred = set()
+
+    def extract_call_layout(args):
+        cur_layout = None
+        for arg in args:
+            if isinstance(arg, _expr.Call):
+                assert arg in _op_layout_map
+                if cur_layout is None:
+                    cur_layout = _op_layout_map[arg]
+                else:
+                    assert cur_layout == _op_layout_map[arg]
+            elif isinstance(arg, _expr.Tuple):
+                return extract_call_layout(arg.fields)
+            elif isinstance(arg, _expr.TupleGetItem):
+                return extract_call_layout(arg.tuple_value.args)
+        return cur_layout
+
+    def visit_func(expr):
+        """Internal visit function"""
+        if isinstance(expr, _expr.Call):
+            cur_layout = None
+            if expr.op == conv2d_op:
+                if expr.attrs.out_layout == "":
+                    _op_layout_map[expr] = expr.attrs.data_layout
+                else:
+                    _op_layout_map[expr] = expr.attrs.out_layout
+                cur_layout = _op_layout_map[expr]
+            else:
+                cur_layout = extract_call_layout(expr.args)
+                if cur_layout is None:
+                    deferred.add(expr)
+                else:
+                    _op_layout_map[expr] = cur_layout
+            if cur_layout is not None:
+                for arg in expr.args:
+                    if arg in deferred:
+                        _op_layout_map[arg] = cur_layout
+                        deferred.remove(arg)
+
+    _analysis.post_order_visit(graph, visit_func)
+    if len(deferred) > 0:
+        raise ValueError
+
+    return _op_layout_map
+
+
+_WEIGHT_SCALE_OPTS = [2**i for i in range(-10, 8)]
+
+
+def slice_idx(begin, end):
+    import tvm.expr
+    assert len(begin) == len(end)
+    for i in range(0, len(begin)):
+        if not isinstance(end[i], tvm.expr.IntImm) or end[i].value - begin[i].value == 0:
+            continue
+        return begin[i].value, end[i].value
+    raise ValueError
+
+
+# ALIGN weight * data scales for convolution
+def match_scales(graph, const_params, mode='max'):
+    conv2d_op = _op.get("nn.conv2d")
+    quantize_op = _op.get("relay.op.annotation.simulated_quantize")
+
+    def visit_func(expr):
+        if isinstance(expr, _expr.Call):
+            if expr.op == conv2d_op:
```

Comment:
when `skip_k_conv > 0`, here will break down. we should skip first `k` conv op.

Reply:
We will remove `skip_k_conv` as it was superseded by a more general `skip_conv_layers`: see #3173